google/cloud/bigquery/_helpers.py

```diff
@@ -133,262 +133,279 @@ def _not_null(value, field):
-        years=years,
-        months=months,
-        days=days,
-        hours=hours,
-        minutes=minutes,
-        seconds=seconds,
-        microseconds=microseconds,
-    )
-
-
-def _float_from_json(value, field):
-    """Coerce 'value' to a float, if set or not nullable."""
-    if _not_null(value, field):
-        return float(value)
-
-
-def _decimal_from_json(value, field):
-    """Coerce 'value' to a Decimal, if set or not nullable."""
-    if _not_null(value, field):
-        return decimal.Decimal(value)
+class CellDataParser:
+    """Converter from BigQuery REST resource to Python value for RowIterator and similar classes.
 
+    See: "rows" field of
+    https://cloud.google.com/bigquery/docs/reference/rest/v2/tabledata/list and
+    https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs/getQueryResults.
+    """
 
-def _bool_from_json(value, field):
-    """Coerce 'value' to a bool, if set or not nullable."""
-    if _not_null(value, field):
-        return value.lower() in ["t", "true", "1"]
+    def to_py(self, resource, field):
+        def default_converter(value, field):
+            _warn_unknown_field_type(field)
+            return value
 
+        converter = getattr(
+            self, f"{field.field_type.lower()}_to_py", default_converter
+        )
+        if field.mode == "REPEATED":
+            return [converter(item["v"], field) for item in resource]
+        else:
+            return converter(resource, field)
+
+    def bool_to_py(self, value, field):
+        """Coerce 'value' to a bool, if set or not nullable."""
+        if _not_null(value, field):
+            # TODO(tswast): Why does _not_null care if the field is NULLABLE or
+            # REQUIRED? Do we actually need such client-side validation?
+            if value is None:
+                raise TypeError(f"got None for required boolean field {field}")
+            return value.lower() in ("t", "true", "1")
+
+    def boolean_to_py(self, value, field):
+        """Coerce 'value' to a bool, if set or not nullable."""
+        return self.bool_to_py(value, field)
+
+    def integer_to_py(self, value, field):
+        """Coerce 'value' to an int, if set or not nullable."""
+        if _not_null(value, field):
+            return int(value)
+
+    def int64_to_py(self, value, field):
+        """Coerce 'value' to an int, if set or not nullable."""
+        return self.integer_to_py(value, field)
+
+    def interval_to_py(
+        self, value: Optional[str], field
+    ) -> Optional[relativedelta.relativedelta]:
+        """Coerce 'value' to an interval, if set or not nullable."""
+        if not _not_null(value, field):
+            return None
+        if value is None:
+            raise TypeError(f"got {value} for REQUIRED field: {repr(field)}")
+
+        parsed = _INTERVAL_PATTERN.match(value)
+        if parsed is None:
+            raise ValueError(f"got interval: '{value}' with unexpected format")
```

Comment:
If there is a URL we can send folks to so that they can get more information on what an acceptable format looks like, please include it in the error message to make this more actionable.

Reply:
That's a good point. https://github.com/googleapis/python-bigquery/pull/2144/commits/3ea3abbc43aa7943564a4a8dd4673311c230cc27 adds URLs for each of the ValueErrors I could find in this PR.